### PR TITLE
Loginlogout + Preference Redux + ProfilePage css

### PR DESCRIPTION
**Pull request by js1998**

### CODE REVIEW

File: imports/redux/actions/index.jsx [hunk 1]
```diff
@@ -25,4 +27,44 @@ export function setMapCenter(coords) {
     coords
   };
 }
-        
\ No newline at end of file
+
+export function loginUser(email) {
```
Comment (hailin-zhang) [inline imports/redux/actions/index.jsx]: We will eventually need password here.

File: imports/redux/reducers/index.jsx [hunk 2]
```diff
@@ -1,26 +1,114 @@
...
+    case ADD_BLACKLIST:
+      let addedBlacklist = state.blacklist.slice();
+      addedBlacklist.push(action.blacklist);
+      return { ...state, blacklist: addedBlacklist}
+    case ADD_FAVOURITES:
+      let addedFavourites = state.favourites.slice();
+      addedFavourites.push(action.favourite);
+      return { ...state, favourites: addedFavourites}
+    case REMOVE_BLACKLIST:
+      let updatedBlacklist = array.filter(function(value, index, array){
```
Comment (hailin-zhang) [inline imports/redux/reducers/index.jsx]: 1. Use `Array.filter`, not `array.filter`
2. remove `function`, change to an arrow function instead

File: imports/redux/reducers/index.jsx [hunk 3]
```diff
@@ -1,26 +1,114 @@
...
+        return action.blacklistToRemove != value;
```
Comment (hailin-zhang) [inline imports/redux/reducers/index.jsx]: !==

File: imports/redux/reducers/index.jsx [hunk 4]
```diff
@@ -1,26 +1,114 @@
...
+  switch (action.type) {
+    case LOGIN_USER:
+    for(let userInfo of userInfos){
+      if(userInfo.email === action.email){
+        return { ...state, blacklist: userInfo.preferences.blacklist, favourites: userInfo.preferences.favourites}
+        break;
+      }
+    }
+    case ADD_BLACKLIST:
+      let addedBlacklist = state.blacklist.slice();
```
Comment (hailin-zhang) [inline imports/redux/reducers/index.jsx]: Can you change these to `const`s?

Comment (hailin-zhang) [inline imports/redux/reducers/index.jsx]: What does the `.slice()` do? Are you just making a shallow copy of the blacklist?

File: imports/redux/reducers/index.jsx [hunk 5]
```diff
@@ -1,26 +1,114 @@
...
+        return action.blacklistToRemove != value;
+      })
+      return { ...state, blacklist: updatedBlacklist}
+    case REMOVE_FAVOURITES:
+      let updatedFavourites = array.filter(function(value, index, array){
+        return action.favouriteToRemove != value;
+      })
+      return { ...state, favourites: updatedFavourites}
+    case LOGOUT_USER:
+      return { ...state, blacklist: [], favourites: []}
```
Comment (hailin-zhang) [inline imports/redux/reducers/index.jsx]: What about the username and email? and `isSignedIn`?

File: imports/redux/reducers/index.jsx [hunk 6]
```diff
@@ -1,26 +1,114 @@
...
+  modalKind: LOGIN,
   radius: 1000,
   initialCenter: {
     lat: 49.263749,
     lng: -123.247480
   },
+  isSignedIn: false,
+  blacklist: [],
+  favourites: [],
+  username: ""
```
Comment (hailin-zhang) [inline imports/redux/reducers/index.jsx]: Thought we weren't using username? We're using full name + email instead?

File: imports/ui/pages/home/components/mapContainer.jsx [hunk 7]
```diff
@@ -52,18 +52,18 @@ export class MapContainer extends React.Component {
             fillColor='#FFB26B'
             fillOpacity={0.4}
           />
-    </Map>   
+    </Map>
     }
 }
 
 const mapStateToProps = (state) => {
   console.log(state.map.initialCenter);
```
Comment (hailin-zhang) [inline imports/ui/pages/home/components/mapContainer.jsx]: Remove this console log please

File: imports/ui/pages/profile/components/blacklist-container.jsx [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+class BlacklistContainer extends React.Component {
+
+    getBlacklist = () => {
+      return Array.from(this.props.blacklist).map((value, index) => {
+                 console.log(value);
+                 return(
+                  <div className="preference-container" key={index}>
+                    <PreferenceButton name={value}/>
+                  </div>
+                )
```
Comment (hailin-zhang) [inline imports/ui/pages/profile/components/blacklist-container.jsx]: Semicolon after round brackets please

File: imports/ui/pages/profile/components/blacklist-container.jsx [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+            }
+        )
```
Comment (hailin-zhang) [inline imports/ui/pages/profile/components/blacklist-container.jsx]: Semicolon

File: imports/ui/pages/profile/components/blacklist-container.jsx [hunk 10]
```diff
@@ -0,0 +1,33 @@
+import React from 'react';
+import '../profile.page.css';
+import PreferenceButton from '../../../shared_components/preferences/preference-button'
+import { connect } from 'react-redux';
+
+class BlacklistContainer extends React.Component {
+
+    getBlacklist = () => {
+      return Array.from(this.props.blacklist).map((value, index) => {
+                 console.log(value);
```
Comment (hailin-zhang) [inline imports/ui/pages/profile/components/blacklist-container.jsx]: Remove console log

File: imports/ui/pages/profile/components/favourites-container.jsx [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+class FavouritesContainer extends React.Component {
+
+    getFavourites = () => {
+      return Array.from(this.props.favourites).map((value, index) => {
+                 return (
+                 <div className="preference-container" key={index}>
+                   <PreferenceButton name={value}/>
+                 </div>);
+            }
+          )
```
Comment (hailin-zhang) [inline imports/ui/pages/profile/components/favourites-container.jsx]: semicolon

File: imports/ui/pages/profile/components/favourites-container.jsx [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    render() {
+        return (<div className="preference">
+          {this.getFavourites()}
+          </div>
+        )
```
Comment (hailin-zhang) [inline imports/ui/pages/profile/components/favourites-container.jsx]: semicolon

File: imports/ui/pages/profile/components/previousTravels.jsx [hunk 13]
```diff
@@ -0,0 +1,12 @@
+import React from 'react';
+import '../profile.page.css';
+
+class PreviousTravels extends React.Component {
+    render() {
+        return <div className="previous-travels">
+        <br/>
```
Comment (hailin-zhang) [inline imports/ui/pages/profile/components/previousTravels.jsx]: never use line breaks

File: imports/ui/pages/profile/components/profile-container.jsx [hunk 14]
```diff
@@ -0,0 +1,16 @@
+import React from 'react';
+import Profile from './profile';
+import PreviousTravels from './previousTravels';
+import '../profile.page.css';
+
+class ProfileContainer extends React.Component {
+    render() {
+        return (
+          <div className="profile-container">
+          <Profile />
```
Comment (hailin-zhang) [inline imports/ui/pages/profile/components/profile-container.jsx]: Fix the indentation

File: imports/ui/pages/routing.page.jsx [hunk 15]
```diff
@@ -11,7 +11,7 @@ function RoutingPage() {
     {/* TODO: enable once pages finished? */}
       <Route path="/about" component={AboutPage} />
       <Route path="/preferences" component={PreferencesPage} />
-      {/* <Route path="/profile" component={ProfilePage} /> */}
+      <Route exact path="/profile" component={ProfilePage} />
```
Comment (hailin-zhang) [inline imports/ui/pages/routing.page.jsx]: remove the exact

File: imports/ui/shared_components/login/login.jsx [hunk 16]
```diff
@@ -12,7 +12,8 @@ import Typography from '@material-ui/core/Typography';
 import { BootstrapButton } from '../MUI/button/bootstrapButton';
 import { withStyles } from '@material-ui/core/styles';
 import { connect } from 'react-redux';
-import { hideModal } from '../../.././redux/actions';
+import { showModal, hideModal, loginUser } from '../../.././redux/actions';
+export const SIGNUP = 1;
```
Comment (hailin-zhang) [inline imports/ui/shared_components/login/login.jsx]: Put this const in redux and import here. See my comment above.

File: imports/ui/shared_components/login/login.jsx [hunk 17]
```diff
@@ -59,11 +60,26 @@ const DialogActions = withStyles((theme) => ({
 
 
 class Login extends React.Component {
+  constructor(props){
+    super(props);
+    this.state = {email: ""};
```
Comment (hailin-zhang) [inline imports/ui/shared_components/login/login.jsx]: Do you need to handle email in state? Can you not just grab the value of the text field when you click the sign up button?

File: imports/ui/shared_components/login/login.jsx [hunk 18]
```diff
@@ -59,11 +60,26 @@ const DialogActions = withStyles((theme) => ({
...
+  }
 
   handleClose = () => {
     this.props.hideModal();
   }
 
+  loginUser = () => {
+    this.props.loginUser(this.state.email);
```
Comment (hailin-zhang) [inline imports/ui/shared_components/login/login.jsx]: Can you not just get `e.target.value` here? Why the extra overhead of setting a state every time the input changes? The view is re-rendered every time state changes. It's many orders of magnitude more expensive to set the state.

File: imports/ui/shared_components/login/login.jsx [hunk 19]
```diff
@@ -94,10 +111,15 @@ class Login extends React.Component {
               type="password"
               fullWidth
             />
+            <DialogContentText id="no-account-yet">
+              <a href="#" onClick={()=>this.props.showModal(SIGNUP)}>
+              Don't have an account yet? Register now!
```
Comment (hailin-zhang) [inline imports/ui/shared_components/login/login.jsx]: Indentation

File: imports/ui/shared_components/login/login.jsx (hunk 16 above)
Comment (hailin-zhang) [reply imports/ui/shared_components/login/login.jsx]: Is this just arbitrarily 1? Why not have it be 0? or 2? We already have constants that are 0-whatever. I'd prefer if every global constant had a different number so there is no confusion.

File: imports/ui/shared_components/login/login.jsx [hunk 20]
```diff
@@ -81,8 +97,9 @@ class Login extends React.Component {
               Enter your details below.
             </DialogContentText>
             <TextField
+              onChange={this.inputEmailChange}
```
Comment (hailin-zhang) [inline imports/ui/shared_components/login/login.jsx]: Is there really a need for this

File: imports/ui/shared_components/login/login.jsx [hunk 21]
```diff
@@ -81,8 +97,9 @@ class Login extends React.Component {
...
               margin="dense"
-              id="name"
+              id="email"
```
Comment (hailin-zhang) [inline imports/ui/shared_components/login/login.jsx]: Do you need the id for anything?

File: imports/ui/shared_components/navbar/navbar.css [hunk 22]
```diff
@@ -151,3 +151,15 @@ TODO: logo
 .MuiTypography-body1, .MuiInputLabel-formControl {
     font-family: 'Rubik', sans-serif !important;
 }
+
+#no-account-yet {
```
Comment (hailin-zhang) [inline imports/ui/shared_components/navbar/navbar.css]: change the name of this class please. Gives no context as to what the fk this is describing.

File: imports/ui/shared_components/signup/signup.jsx [hunk 23]
```diff
@@ -0,0 +1,135 @@
...
+
+class Signup extends React.Component {
+
+  handleClose = () => {
+    this.props.hideModal();
+  }
+
+  render() {
+    return (
+      <div>
```
Comment (hailin-zhang) [inline imports/ui/shared_components/signup/signup.jsx]: Why is there a div here? Looks like `<Dialog>` already encapsulates everything.

File: imports/redux/reducers/index.jsx (hunk 6 above)
Comment (hailin-zhang) [reply imports/redux/reducers/index.jsx]: Having a defined username string eliminates the need for `isSignedIn`

File: imports/redux/reducers/index.jsx [hunk 24]
```diff
@@ -1,26 +1,114 @@
...
+    case LOGOUT_USER:
+      return { ...state, isSignedIn: false, username: "", email: ""};
+    default:
+      return state;
+  }
+}
+
+function preferencesReducer(state = initialState, action){
+  switch (action.type) {
+    case LOGIN_USER:
```
Comment (hailin-zhang) [inline imports/redux/reducers/index.jsx]: Remove this case. You already defined LOGIN_USER in `loginReducer`.

File: imports/redux/reducers/index.jsx [hunk 25]
```diff
@@ -1,26 +1,114 @@
...
   }
 }
 
+function loginReducer(state = initialState, action) {
+  switch (action.type) {
+    case LOGIN_USER:
+      for(let userInfo of userInfos){
+        if(userInfo.email === action.email){
+          return { ...state, email: action.email, isSignedIn: true, username: `${userInfo.firstName} ${userInfo.lastName}`}
+          break;
```
Comment (hailin-zhang) [inline imports/redux/reducers/index.jsx]: This is unreachable code. Please remove

File: imports/redux/reducers/index.jsx [hunk 26]
```diff
@@ -1,26 +1,114 @@
...
+        }
+      }
+      alert("Invalid login info. Try Again.");
+      break;
```
Comment (hailin-zhang) [inline imports/redux/reducers/index.jsx]: We're going to refactor this reducer completely eventually, can you make a note of this? Add a TODO

File: imports/redux/reducers/index.jsx [hunk 27]
```diff
@@ -1,26 +1,114 @@
...
+      let addedFavourites = state.favourites.slice();
+      addedFavourites.push(action.favourite);
+      return { ...state, favourites: addedFavourites}
+    case REMOVE_BLACKLIST:
+      let updatedBlacklist = array.filter(function(value, index, array){
+        return action.blacklistToRemove != value;
+      })
+      return { ...state, blacklist: updatedBlacklist}
+    case REMOVE_FAVOURITES:
+      let updatedFavourites = array.filter(function(value, index, array){
```
Comment (hailin-zhang) [inline imports/redux/reducers/index.jsx]: `Array.filter`, use arrow functions

File: imports/redux/reducers/index.jsx [hunk 28]
```diff
@@ -1,26 +1,114 @@
...
+    }
+    case ADD_BLACKLIST:
+      let addedBlacklist = state.blacklist.slice();
+      addedBlacklist.push(action.blacklist);
+      return { ...state, blacklist: addedBlacklist}
+    case ADD_FAVOURITES:
+      let addedFavourites = state.favourites.slice();
+      addedFavourites.push(action.favourite);
+      return { ...state, favourites: addedFavourites}
+    case REMOVE_BLACKLIST:
```
Comment (hailin-zhang) [inline imports/redux/reducers/index.jsx]: Can you keep the blacklist cases together? and the favourite cases together?

File: imports/redux/reducers/index.jsx [hunk 29]
```diff
@@ -1,26 +1,114 @@
...
+      let updatedBlacklist = array.filter(function(value, index, array){
+        return action.blacklistToRemove != value;
+      })
+      return { ...state, blacklist: updatedBlacklist}
+    case REMOVE_FAVOURITES:
+      let updatedFavourites = array.filter(function(value, index, array){
+        return action.favouriteToRemove != value;
+      })
+      return { ...state, favourites: updatedFavourites}
+    case LOGOUT_USER:
```
Comment (hailin-zhang) [inline imports/redux/reducers/index.jsx]: Remove this, you already defined this case in `loginReducer`

File: imports/ui/pages/profile/components/blacklist-container.jsx [hunk 30]
```diff
@@ -0,0 +1,33 @@
...
+                    <PreferenceButton name={value}/>
+                  </div>
+                )
+            }
+        )
+    }
+
+    render() {
+        return <div className="preference">
+          {this.getBlacklist()}
```
Comment (hailin-zhang) [inline imports/ui/pages/profile/components/blacklist-container.jsx]: Why are there `()`s here?

File: imports/ui/pages/profile/components/blacklist-container.jsx [hunk 31]
```diff
@@ -0,0 +1,33 @@
...
+                  <div className="preference-container" key={index}>
+                    <PreferenceButton name={value}/>
+                  </div>
+                )
+            }
+        )
+    }
+
+    render() {
+        return <div className="preference">
```
Comment (hailin-zhang) [inline imports/ui/pages/profile/components/blacklist-container.jsx]: more descriptive class names please. `preferences` gives me no clue as to what the css is being applied to.

File: imports/ui/pages/profile/components/favourites-container.jsx [hunk 32]
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+import '../profile.page.css';
+import PreferenceButton from '../../../shared_components/preferences/preference-button'
+import { connect } from 'react-redux';
+
+class FavouritesContainer extends React.Component {
+
+    getFavourites = () => {
+      return Array.from(this.props.favourites).map((value, index) => {
+                 return (
```
Comment (hailin-zhang) [inline imports/ui/pages/profile/components/favourites-container.jsx]: fix indentation

File: imports/redux/reducers/index.jsx [hunk 33]
```diff
@@ -1,26 +1,114 @@
 import { combineReducers } from 'redux';
-import { SHOW_MODAL, HIDE_MODAL, SET_RADIUS, SET_CENTER } from '../actions/index';
+import { SHOW_MODAL, HIDE_MODAL, SET_RADIUS, SET_CENTER , LOGIN_USER, LOGOUT_USER,
```
[the remark this answers is not in the document]
Comment (js1998) [reply imports/redux/reducers/index.jsx]: What do you mean by "line-by-line, comma-separated"?

[the remark this answers is not in the document]
Comment (hailin-zhang) [reply imports/redux/reducers/index.jsx]:
```
import {
SHOW_MODAL,
HIDE_MODAL,
SET_RADIUS,
...
} from ...
```

File: imports/redux/reducers/index.jsx [hunk 34]
```diff
@@ -1,26 +1,114 @@
...
+  GET_PREFERENCES, ADD_BLACKLIST, ADD_FAVOURITES, REMOVE_BLACKLIST, REMOVE_FAVOURITES } from '../actions/index';
+import { LOGIN, SIGNUP } from '../../ui/shared_components/navbar/navbar';
+
+let userInfos = [
```
[the remark this answers is not in the document]
Comment (jvssicawu) [reply imports/redux/reducers/index.jsx]: what do u mean by add a TODO

[the remark this answers is not in the document]
Comment (jvssicawu) [reply imports/redux/reducers/index.jsx]: oh nvm gotcha

File: imports/ui/shared_components/login/login.jsx (hunk 21 above)
Comment (js1998) [reply imports/ui/shared_components/login/login.jsx]: nope, i can remove it

**Pull request merged**